kie-server-parent/kie-server-services/kie-server-services-optaplanner/src/main/java/org/kie/server/services/optaplanner/OptaplannerKieServerExtension.java

```diff
@@ -105,10 +112,28 @@ public void destroy(KieServerImpl kieServer, KieServerRegistry registry) {
         if( this.threadPool != null ) {
             this.threadPool.shutdownNow();
         }
+        
+        for (ContinuousPlanning planning : globalContinuousPlannings) {
+            try {
+                planning.close();
+                logger.debug("ContinuousPlanning {} closed successfully", planning);
+            } catch (Exception e) {
+                logger.error("Unexpected error thrown when closing {}", planning, e);
+            }
+        }
     }
 
     @Override
     public void createContainer(String id, KieContainerInstance kieContainerInstance, Map<String, Object> parameters) {
+        
+        List<ContinuousPlanning> containerPlannings = loadPlannings(kieContainerInstance.getKieContainer().getClassLoader());
+        containerContinuousPlannings.put(id, containerPlannings);
+        // first try with kjars defined plannings
+        boolean found = solverServiceBase.boostrapContinuousPlanning(id, containerPlannings);
+        if (!found) {
+            // if not found then use global ones
+            solverServiceBase.boostrapContinuousPlanning(id, globalContinuousPlannings);
```

Comment:
Just to be sure - so global ones will always return true for their accept method, right?

Reply:
no, but it does not matter anymore as there is nothing more we can do so it's essentially done regardless if it found anything or not.

After I did this I think that global are actually not needed, reason for that is ProjectClassLoader has kie server app class loader as parent anyway so that should discover globals too.

Reply:
OK, but what is the difference between container and global continuous plannings? I mean container plannings are connected with a particular container - therefore they are in a map. However, global ones are not connected with any container so why are they loaded using the same method when they are not connected to any container? But maybe I am just missing something here...

Reply:
OK, I might understand it now - so the only difference is the classloader but still it is needed for each planning to have its accepted container on which it can run, right? :)

Reply:
the reason here was to have single method that can load and init planning from different class loaders. globals were for these that are on kie server class path while the containers are for those that come with kjar so they need to be additionally tracked so can be closed the kjar is undeployed

Reply:
ok clear now. So if there aren't any container-specific plannings for the container being deployed then with each container deployment the global planning (if it is accepted for the container being deployed, let's say it always returns true) will be started, right? So theoretically if I set up it this way it will act as a fallback for containers that don't have their own container-specific planning.

Reply:
correct

Reply:
@mswiderski is it possible that due to the fact that "ProjectClassLoader has kie server app class loader as parent" might be some "global" planning loaded twice? Firstly as global planning, secondly as container-specific?

Reply:
@rsynek yes, this is what I mentioned in the comment above that ProjectCLassLoader will find the global planning too. So most likely there is no need for them (global plannings) to be loaded separately. and yes, that will result in multiple loads in case there are several containers and planning on kie server class path.

Reply:
@mswiderski my concern is whether it will behave as you described - it seems that container-specific is not overriding the global planning, as in fact it's going to load both due to the classloader hierarchy.

Reply:
Ok, thanks for explanation